File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        self.answer = []
        self.n = len(nums)

        self.generate(nums, [], [False] * self.n)

        return self.answer

    def generate(self, nums, current, vis):
        if len(current) == self.n:
            self.answer.append(current.copy())
            return

        for i in range(self.n):
            if vis[i]:
                continue

            vis[i] = True
            current.append(nums[i])

            self.generate(nums, current, vis)

            vis[i] = False
            current.pop()

File: 0046-permutations/0046-permutations.cpp
class Solution {
    vector<vector<int>> answer;
    int n;
public:
    vector<vector<int>> permute(vector<int>& nums) {
        n = nums.size();
        generate(nums,vector<int>(),vector<bool>(n,false));
        return answer;
    }
    void generate(vector<int>& nums,vector<int> current,vector<bool> vis) {
        if (current.size() == n) {
            answer.push_back(current);
            return;
        }
        for (int i = 0;i<n;i++) {
            if (vis[i]) continue;
            vis[i] = true;
            current.push_back(nums[i]);
            generate(nums,current,vis);
            vis[i] = false;
            current.pop_back();
        }
    }
};